docs/proposals/210701_endpoint_discovery.md

```diff
@@ -0,0 +1,44 @@
+---
+title: Unified Endpoint discovery
+type: proposal
+menu: proposals
+status: XXX
+---
+
+### Related Tickets
+
+* https://github.com/thanos-io/thanos/issues/2600
+
+### Summary
+
+We want to propose a new flag called `--endpoint=<address>` that will be passed to Thanos query and replace the current `--store=<address>` and `--rule=<address>` flags, and add a discovery mechanism which will be used to determine what types of services this endpoint can serve.
+
+### Motivation
+
+Currently, in Thanos Query, the discovery of rules APIs happens via Store API's Info method, this makes it harder if we ever want to not have a coupling to the Store API (which is already planned for [scalable ruler proposal](https://github.com/thanos-io/thanos/blob/master/docs/proposals/202005_scalable-rule-storage.md)).
+
+We also require passing two different flags to the Thanos Query component `--store=<address>` and `--rule=<address>` and perform DNS discovery on often the same address multiple times, this can cause occasional DNS issues. Which is especially confusing when one works and the other doesn't.
+
+### Goals
+
+Unite discovery of endpoints, which would make code cleaner as well as provide building blocks to provide discover and support for new gRPC services, for example, the Targets API/Exemplar API/Metadata API.
+
+### Solution
+
+Add a new flag called `--endpoint` to Thanos query, and auto-discover what services that endpoint is serving based on metadata each gRPC server exposes.
+
+The plan would be for the initial discovery of gRPC services available to happen via gRPC reflection. To get further information, we would be adding an Info method to the Rules API. This would contain similar info or metadata information as we get right now from the Store Info method. Potentially in the future, we would implement this same Info method for other services in the, for example, the Targets, Examplers, etc. As mentioned the discovery would happen via reflection initially, the Info method in the Rules API would expose extra metadata about external labels for now, and we can always add more data to it in the future as needed.
```

Comment:
Could you link to a page that explains gRPC reflection in a bit more detail? I'm not too familiar and think most others might be neither.

Reply:
Of course, good point! Would linking to [this](https://github.com/grpc/grpc/blob/master/doc/server-reflection.md) be enough, or also the various functions we would actually call to perform it? I have done the POC locally so can include those as well, not sure if its too much details then?